netket/nn/symmetric_linear.py

```diff
@@ -101,62 +100,407 @@ def full_bias(self, bias):
+        dtype = jnp.promote_types(x.dtype, self.dtype)
+        x = jnp.asarray(x, dtype)
 
         kernel = self.param(
-            "kernel", self.kernel_init, (inputs.shape[-1], self.features), self.dtype
+            "kernel", self.kernel_init, (x.shape[-1], self.features), self.dtype
         )
-        kernel = self.full_kernel(kernel)
+
+        if self.mask:
+            kernel = kernel * jnp.expand_dims(self.mask, 1)
+
+        kernel = self.full_kernel(kernel).reshape(-1, self.features, self.n_symm)
         kernel = jnp.asarray(kernel, dtype)
 
-        y = lax.dot_general(
-            inputs,
+        x = lax.dot_general(
+            x,
             kernel,
-            (((inputs.ndim - 1,), (0,)), ((), ())),
+            (((x.ndim - 1,), (0,)), ((), ())),
             precision=self.precision,
         )
 
         if self.use_bias:
             bias = self.param("bias", self.bias_init, (self.features,), self.dtype)
             bias = jnp.asarray(self.full_bias(bias), dtype)
-            y += bias
+            x += bias
 
-        return y
+        return x
 
 
-class DenseEquivariant(Module):
-    r"""A group convolution operation that is equivariant over a symmetry group
+class DenseSymmFFT(Module):
+    """Implements a symmetrized projection onto a space group using a Fast Fourier Transform"""
 
-    Acts on a feature map of symmetry poses of shape [batch_size,n_symm*in_features]
-    and returns a feature  map of poses of shape [batch_size,n_symm*out_features]
+    space_group: HashableArray
+    """Array that lists the space group as permutations"""
+    features: int
+    """The number of input features; must be the second dimension of the input."""
+    shape: Tuple
+    """Tuple that corresponds to shape of lattice"""
+    use_bias: bool = True
+    """Whether to add a bias to the output (default: True)."""
+    mask: Optional[HashableArray] = None
+    """Mask that zeros out elements of the filter. Should be of shape [inp.shape[-1]]"""
+    dtype: DType = jnp.float64
+    """The dtype of the weights."""
+    precision: Any = None
 
-    G-convolutions are described in ` Cohen et. {\it al} <http://proceedings.mlr.press/v48/cohenc16.pdf>`_
-    and applied to quantum many-body problems in ` Roth et. {\it al} <https://arxiv.org/pdf/2104.05085.pdf>`_
+    kernel_init: NNInitFunc = unit_normal_scaling
+    """Initializer for the Dense layer matrix. Defaults to variance scaling"""
+    bias_init: NNInitFunc = zeros
+    """Initializer for the bias. Defaults to zero initialization"""
 
-    The G-convolution generalizes the convolution to non-commuting groups:
+    def setup(self):
+        self.n_cells = np.product(np.asarray(self.shape))
+        self.n_symm = len(np.asarray(self.space_group)) // self.n_cells
+        self.sites_per_cell = np.asarray(self.space_group).shape[1] // self.n_cells
+
+        self.mapping = (
+            np.asarray(self.space_group)[:, : self.sites_per_cell]
```

Comment:
It would be more elegant to define a variable for `np.asarray(self.space_group)` both here and in `DenseEquivariant`. It shouldn't come with a memory overhead, right? In any case it wouldn't be more than doing it this way

Reply:
Yeah I can do that